File: two_pointer/triplet-sum-to-zero.js
const search_triplets = (arr) => {
    let result = [];
    arr.sort((a, b) => a - b);

    for (let i = 0; i < arr.length; i++) {
        let l = i + 1;
        let r = arr.length - 1;
        let target = 0 - arr[i];

        while (l < r && arr[l] !== arr[l - 1]) {
            arr[l] + arr[r] === target && result.push([arr[i], arr[l], arr[r]]);
            (arr[l] + arr[r] > target) ? r-- : l++;
        }
    }

    return result;
}

console.log(search_triplets([-3, 0, 1, 2, -1, 1, -2])) // [[-3, 1, 2], [-2, 0, 2], [-2, 1, 1], [-1, 0, 1]]
console.log(search_triplets([-5, 2, -1, -2, 3])) // [[-5, 2, 3], [-2, -1, 3]]

File: two_pointer/python/triplet-sum-to-zero.py
def search_triplets(arr):
    triplets = []

    arr.sort()

    i = 0
    while i < len(arr) - 2:
        j = arr[i]
        target_sum = 0 - j
        l, r = i + 1, len(arr) - 1
        while l < r:
            current_sum = arr[l] + arr[r]
            if current_sum == target_sum:
                while l == l + 1:
                    l += 1
                while r == r - 1:
                    r -= 1
                triplets.append([j, arr[l], arr[r]])
                l += 1
                r -= 1

            elif current_sum > target_sum:
                r -= 1
            else:
                l += 1
        i += 1

    return triplets


if __name__ == '__main__':
    print(search_triplets([-3, 0, 1, 2, -1, 1, -2]))
    print(search_triplets([-5, 2, -1, -2, 3]))
